tests/scanner/scanners/external_project_access_scanner_test.py

```diff
@@ -0,0 +1,301 @@
+
+from google.cloud.forseti.common.gcp_api import cloud_resource_manager as crm
+from google.cloud.forseti.common.gcp_api import api_helpers
+from google.cloud.forseti.common.gcp_type import resource_util
+from google.cloud.forseti.common.gcp_type.folder import Folder
+from google.cloud.forseti.common.gcp_type.organization import Organization
+from google.cloud.forseti.common.gcp_type.project import Project
+from google.cloud.forseti.common.gcp_type import resource as resource_mod
+from google.cloud.forseti.common.util import logger
+from google.cloud.forseti.scanner.audit.external_project_access_rules_engine import Rule
+from google.cloud.forseti.scanner.scanners import external_project_access_scanner as epas
+from google.cloud.forseti.services.inventory.storage import initialize
+from google.cloud.forseti.services.inventory.storage import Storage
+from tests.common.gcp_api.test_data import fake_crm_responses
+from tests.common.gcp_api.test_data import http_mocks
+from tests.services.api_tests.inventory_test import TestServiceConfig
+from tests.services.inventory.storage_test import ResourceMock
+from tests.services.util.db import create_test_engine
+from tests.unittest_utils import ForsetiTestCase
+from tests.unittest_utils import get_datafile_path
+# pylint: enable=line-too-long
+
+TEST_EMAILS = ['user1@example.com', 'user2@example.com']
+
+LOGGER = logger.get_logger(__name__)
+
+
+class ExternalProjectAccessScannerTest(ForsetiTestCase):
+    """Tests for ExternalProjectAccessScanner"""
+
+    def setUp(self):
+        """Setup"""
+        self.global_configs = dict()
+        self.scanner_configs = (
+            dict(output_path='gs://test-forseti-dev/scanner/output',
+                 rules_path=__file__,
+                 scanners=[dict(name='external_project_access', enabled=True)]))
+
+        mock_invnetory_config = mock.MagicMock()
+        mock_invnetory_config.get_root_resource_id.return_value = (
+            'organizations/567890')
+
+        mock_service_config = mock.MagicMock()
+        mock_service_config.get_inventory_config.return_value = (
+            mock_invnetory_config)
+
+        self.service_config = mock_service_config
+        self.model_name = 'TestModel'
+        self.snapshot_timestamp = datetime.now().strftime('%Y%m%dT%H%M%SZ')
+        self.rules = get_datafile_path(
+            __file__, 'external_project_access_test_rules_1.yaml')
+
+        self.stash_email_method = epas.get_user_emails
+
+    def test_mock_emails(self):
+        """Test obtaining e-mails from mock"""
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        self.assertEqual(len(epas.get_user_emails(None)), 2)
+
+    #pylint: disable=W0212
+    def test_retrieve(self):
+        """Test retrieving project ancestry data"""
+        list_to_generator = lambda x: (n for n in x)
+
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        epas.extract_project_ids = mock.MagicMock(
```

Comment:
I'd like to see this function tested without mocking. Can you mock the crm_client instead, to return mock projects?

Reply:
We will take care of this comments in a follow-up PR.